### PR TITLE
NXP-30210: Fix variable interpolation in pipelines (branch 21)

### PR DESCRIPTION
**Pull request by atchertchian**
Visible warnings:

- build:
        sh: [CONNECT_PASS]
        sh: [EXPLORER_PASSWORD, UNUSED_USERNAME]
- release:
        sh: [CONNECT_PASS]
- export:
        sh: [INSTANCE_CLID]
        sh: [EXPLORER_PASS]

### CODE REVIEW

**Review by kevinleturc — APPROVED** [review]

File: ci/Jenkinsfiles/export.groovy [hunk 1]
```diff
@@ -183,14 +183,14 @@ pipeline {
             script {
               def moduleDir = 'docker/nuxeo-explorer-export-docker'
               def oneLineClid = getClid("${INSTANCE_CLID}")
-              sh """#!/bin/bash +x
-                CONNECT_EXPLORER_CLID="${oneLineClid}" \
-                CONNECT_EXPORT_CLID="${oneLineClid}" \
+              sh '''#!/bin/bash +x
+                CONNECT_EXPLORER_CLID=$oneLineClid \
+                CONNECT_EXPORT_CLID=$oneLineClid \
```
Comment (ataillefer) [inline ci/Jenkinsfiles/export.groovy:188]: Not sure about these 2 as `oneLineClid` is a Groovy variable, it probably needs to be interpolated.
Keeping the existing code should work and not trigger the security warning.

File: ci/Jenkinsfiles/build.groovy [hunk 2]
```diff
@@ -320,12 +320,11 @@ pipeline {
               boolean isReferenceBranch = BRANCH_NAME == REFERENCE_BRANCH
               // first substitute docker image names and versions
               withCredentials([usernamePassword(credentialsId: 'explorer-preview-maintenance', passwordVariable: 'EXPLORER_PASSWORD', usernameVariable: 'UNUSED_USERNAME')]) {
-                def explorerPasswordProp = "org.nuxeo.apidoc.apidocAdmin.password=${EXPLORER_PASSWORD}"
-                sh """
+                sh '''
                   mv values.yaml values.yaml.tosubst
-                  NUXEO_EXPLORER_CUSTOM_PARAMS=${isReferenceBranch ? explorerPasswordProp : ''} \
+                  NUXEO_EXPLORER_CUSTOM_PARAMS=${isReferenceBranch ? 'org.nuxeo.apidoc.apidocAdmin.password=$EXPLORER_PASSWORD' : ''} \
```
Comment (atchertchian) [inline ci/Jenkinsfiles/build.groovy:325]: This one fails on master with:
```
line 3: ${isReferenceBranch ? 'org.nuxeo.apidoc.****.password=$EXPLORER_PASSWORD' : ''}: bad substitution
```

**Comment by atchertchian** [comment]
Reverted the commit on master because it makes preview fail

File: ci/Jenkinsfiles/build.groovy (hunk 2 above)
Comment (ataillefer) [reply ci/Jenkinsfiles/build.groovy:325]: Yes, same issue as https://github.com/nuxeo/nuxeo-explorer/pull/105/files#r584718096 probably.
I think you can try `"org.nuxeo.apidoc.apidocAdmin.password=${EXPLORER_PASSWORD}"` to have the substitution.

File: ci/Jenkinsfiles/export.groovy (hunk 1 above)
Comment (atchertchian) [reply ci/Jenkinsfiles/export.groovy:188]: I do get a warning on [INSTANCE_CLID], so maybe it's about changing
```
def oneLineClid = getClid("${INSTANCE_CLID}")
```
into 
```
def oneLineClid = getClid($INSTANCE_CLID)
```

?

File: ci/Jenkinsfiles/build.groovy (hunk 2 above)
Comment (atchertchian) [reply ci/Jenkinsfiles/build.groovy:325]: Actually sorry i do have a warn on EXPLORER_PASSWORD so maybe i should do the following change:
Replace
```
def explorerPasswordProp = "org.nuxeo.apidoc.apidocAdmin.password=${EXPLORER_PASSWORD}"
```
By
```
def explorerPasswordProp = "org.nuxeo.apidoc.apidocAdmin.password=$EXPLORER_PASSWORD"
```

?

Comment (ataillefer) [reply ci/Jenkinsfiles/build.groovy:325]: We want to get rid of Groovy interpolation, aka "${SENSITIVE NEV VAR}" or """ ${SENSITIVE NEV VAR} """, meaning, we cannot have sensitive environment variables manipulated in a Groovy script...

So, in this case, I guess the simpler would be:
```
          sh 'mv values.yaml values.yaml.tosubst'
          if (isReferenceBranch) {
            withCredentials([usernamePassword(credentialsId: 'explorer-preview-maintenance', passwordVariable: 'EXPLORER_PASSWORD', usernameVariable: 'UNUSED_USERNAME')]) {
              sh '''
                NUXEO_EXPLORER_CUSTOM_PARAMS=org.nuxeo.apidoc.apidocAdmin.password=$EXPLORER_PASSWORD \
                  envsubst < values.yaml.tosubst > values.yaml
              '''
            }
          } else {
            sh 'envsubst < values.yaml.tosubst > values.yaml'
          }
```

File: ci/Jenkinsfiles/export.groovy (hunk 1 above)
Comment (atchertchian) [reply ci/Jenkinsfiles/export.groovy:188]: Alright thanks :+1: 

Aren't the double quotes needed for proper resolution of `${moduleDir}` ?

Comment (ataillefer) [reply ci/Jenkinsfiles/export.groovy:188]: yes!

**Review by ataillefer — APPROVED** [review]

**Comment by atchertchian** [comment]
(fixed misplaced script in export pipeline, see 21e439ae8fb989a07d01df3d2d5a74b16f676232)

**Pull request merged**